sage.py

```diff
@@ -571,243 +485,159 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-    interesting = []
-    # Reorganize data for each attacker per team 
+
+    # Reorganize data for each attacker per team
     team_data = dict()
-    s_t = dict()
-    for tid,team in enumerate(unparse):
-        #attackers = list(set([x[1] for x in team])) # collecting all src ip
-        #attackers.extend(list(set([x[3] for x in team]))) # collection all dst ip
-        #attackers = [x for x in attackers if x not in hostip.keys()] # filtering only attackers
-        
-        
-        host_alerts = dict()
-
-        for ev in team:
-            #print(ev[0])
-            h = ev[7]
-            s = ev[1]
-            d = ev[3]
-            c = ev[9]
-            ts = ev[8]
-            sp = ev[2] if ev[2] != None else 65000 
-            dp = ev[4] if ev[4] != None else 65000
-            signature = ev[5]
-            # Simply respect the source,dst format! (Correction: source is always source and dest alwyas dest!)
-            
-            source, dest, port = -1, -1, -1
-            #print(s, d, sp, dp)
-            #assert sp >= dp
-            
-            source = s #if s not in inv_hostip.keys() else inv_hostip[s]
-            dest = d #if d not in inv_hostip.keys() else inv_hostip[d]
-            # explicit name if cant resolve
-            #port = str(dp) if (dp not in port_services.keys() or port_services[dp] == 'unknown') else port_services[dp]['name'] 
-            
-            # say unknown if cant resolve it
-            port = 'unknown' if (dp not in port_services.keys() or port_services[dp] == 'unknown') else port_services[dp]['name']
+    for tid, team in enumerate(team_alerts):
+        host_alerts = dict()  # (attacker, victim) -> alerts
+
+        for alert in team:
+            # Alert format: (diff_dt, src_ip, src_port, dst_ip, dst_port, sig, cat, host, ts, mcat)
+            src_ip, dst_ip, signature, ts, mcat = alert[1], alert[3], alert[5], alert[8], alert[9]
+            dst_port = alert[4] if alert[4] is not None else 65000
+            # Simply respect the source,dst format! (Correction: source is always source and dest always dest!)
+
+            # Say 'unknown' if the port cannot be resolved
+            dst_port = 'unknown' if (dst_port not in port_services.keys() or port_services[dst_port] == 'unknown') else port_services[dst_port]['name']
             
-            if (source,dest) not in host_alerts.keys() and (dest,source) not in host_alerts.keys():
-                host_alerts[(source,dest)] = []
-                #print(tid, (source,dest), 'first', ev[8])
-                s_t[str(tid)+"|"+str(source)+"->"+str(dest)] = ev[8]
-                
-            if((source,dest) in host_alerts.keys()):
-                host_alerts[(source,dest)].append((dest, c, ts, port, signature)) # TODO: remove the redundant host names
-                #print(source, dest, (micro[c].split('.'))[-1], port)
+            if (src_ip, dst_ip) not in host_alerts.keys() and (dst_ip, src_ip) not in host_alerts.keys():
```

Comment:
These lines. Should they be kept? `(src_ip, dst_ip)` is added to `host_alerts` in either case

Reply:
Yes. The first if() simply adds an empty list and creates the dict key. The second if,else() actually append the tuples in the respective list.